internal/appsec/emitter/graphqlsec/execution.go

```diff
@@ -0,0 +1,86 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016 Datadog, Inc.
+
+// Package graphql is the GraphQL instrumentation API and contract for AppSec
+// defining an abstract run-time representation of AppSec middleware. GraphQL
+// integrations must use this package to enable AppSec features for GraphQL,
+// which listens to this package's operation events.
+package graphqlsec
+
+import (
+	"context"
+	"reflect"
+
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/appsec/dyngo"
+	"gopkg.in/DataDog/dd-trace-go.v1/internal/appsec/trace"
+)
+
+type Execution struct {
+	dyngo.Operation
+	trace.TagSetter
+	trace.SecurityEventsHolder
+}
+
+// ExecutionArguments describes arguments passed to a GraphQL query operation.
+type ExecutionArguments struct {
+	// Variables is the user-provided variables object for the query.
+	Variables map[string]any
+	// Query is the query that is being executed.
+	Query string
+	// OperationName is the user-provided operation name for the query.
+	OperationName string
+}
+
+// StartExecution starts a new GraphQL query operation, along with the given arguments, and emits a
+// start event up in the operation stack.
+func StartExecution(ctx context.Context, span trace.TagSetter, args ExecutionArguments, listeners ...dyngo.DataListener) (context.Context, *Execution) {
+	// The parent will typically be the Request operation that previously fired...
+	parent, _ := ctx.Value(contextKey{}).(dyngo.Operation)
```

Comment:
I wonder if not having a parent operation present should be considered a valid use of graphqlsec or not.

If your definition supposes that Request > Execution > Field are nested together properly, then I'd suggest you enforce it (as a strongly typed `parent` argument or by type-asserting it here?).

I quickly checked what was done in grpcsec, and it is only enforced by construction (the grpc middleware is a huge closure with all the dyngo operations available so it's simple to pass them to new children operations).

Reply:
Fixed in b737b29326754c55def646d0e0d30f75b8199cf0